samples/bluetooth/hci_rpmsg/src/main.c

```diff
@@ -0,0 +1,455 @@
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	.regions = {
+		{
+			.virt       = (void *) SHM_START_ADDR,
+			.physmap    = shm_physmap,
+			.size       = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask  = 0xffffffff,
+			.mem_flags  = 0,
+			.ops        = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static struct virtqueue *vq[2];
+static struct rpmsg_endpoint ep;
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return sys_read8(VDEV_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return BIT(VIRTIO_RPMSG_F_NS);
+}
+
+static void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	sys_write8(status, VDEV_STATUS_ADDR);
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	status = ipm_send(ipm_tx_handle, 0, 0, NULL, 0);
+	if (status != 0) {
+		LOG_ERR("ipm_send failed to notify: %d", status);
+	}
+}
+
+const struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.set_status = virtio_set_status,
+	.get_features = virtio_get_features,
+	.notify = virtio_notify,
+};
+
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	LOG_INF("Got callback of id %u", id);
+	virtqueue_notification(vq[1]);
+}
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+static K_THREAD_STACK_DEFINE(tx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread tx_thread_data;
+
+/* HCI command buffers */
+#define CMD_BUF_SIZE BT_BUF_RX_SIZE
+NET_BUF_POOL_DEFINE(cmd_tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+#if defined(CONFIG_BT_CTLR_TX_BUFFER_SIZE)
+#define BT_L2CAP_MTU (CONFIG_BT_CTLR_TX_BUFFER_SIZE - BT_L2CAP_HDR_SIZE)
+#else
+#define BT_L2CAP_MTU 65 /* 64-byte public key + opcode */
+#endif /* CONFIG_BT_CTLR */
+
+/** Data size needed for ACL buffers */
+#define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
+
+#if defined(CONFIG_BT_CTLR_TX_BUFFERS)
+#define TX_BUF_COUNT CONFIG_BT_CTLR_TX_BUFFERS
+#else
+#define TX_BUF_COUNT 6
+#endif
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, TX_BUF_COUNT, BT_BUF_ACL_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+static K_FIFO_DEFINE(tx_queue);
+
+#define HCI_RPMSG_CMD 0x01
+#define HCI_RPMSG_ACL 0x02
+#define HCI_RPMSG_SCO 0x03
+#define HCI_RPMSG_EVT 0x04
+
+static struct net_buf *hci_rpmsg_cmd_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_cmd_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		LOG_ERR("Not enought data for command header");
+		return NULL;
+	}
+
+	buf = net_buf_alloc(&cmd_tx_pool, K_NO_WAIT);
+	if (buf) {
+		bt_buf_set_type(buf, BT_BUF_CMD);
+
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		LOG_ERR("No available command buffers!");
```

Comment:
Don't you need to `return NULL;` in this branch? Further down the function there would otherwise be NULL-pointer dereferences, like the `net_buf_add_mem()` call.

Reply:
fixed